File: decl-py/tests/render_test.py
"""The render corpus (tests/render) through the native command line: the
format goldens of formats.json (`--format yaml`, `--indent n`, and the
YAML read back to the golden document), every golden document bound from
its YAML twin under inputs/, and the documents under invalid/ that the
reader must refuse with their messages (tests/render/README.md)."""

from __future__ import annotations

import json
import re
import subprocess
from pathlib import Path
from typing import Any

import pytest

from decl.semantics import read_json
from decl.yaml import YamlError, read_yaml, to_json

ROOT = Path(__file__).resolve().parents[2]
MANIFEST = json.loads((ROOT / "tests/golden/manifest.json").read_text(encoding="utf-8"))
FORMATS = json.loads((ROOT / "tests/render/formats.json").read_text(encoding="utf-8"))
INVALID = json.loads((ROOT / "tests/render/invalid/cases.json").read_text(encoding="utf-8"))


def module_of(entry: dict[str, Any], tmp_path: Path) -> str:
    """a markdown entry's module: its ```decl blocks in order, in a temporary file"""
    if "module" in entry:
        return str(entry["module"])
    md = (ROOT / entry["markdown"]).read_text(encoding="utf-8")
    src = "\n".join(re.findall(r"```decl\n([\s\S]*?)```", md))
    p = tmp_path / "guide.decl"
    p.write_text(src, encoding="utf-8")
    return str(p)


def args_of(entry: dict[str, Any], tmp_path: Path, inputs: list[str] | None = None) -> list[str]:
    args = ["validate" if entry.get("rejected") else "evaluate", module_of(entry, tmp_path)]
    for spec in entry.get("inputs", []) if inputs is None else inputs:
        args += ["--input", spec]
    if "output" in entry:
        args += ["--output", entry["output"]]
    return args


def run(python: str, args: list[str]) -> subprocess.CompletedProcess[str]:
    return subprocess.run([python, "-m", "decl", *args], cwd=ROOT, capture_output=True, text=True)


@pytest.mark.parametrize("f", FORMATS, ids=[f["golden"] for f in FORMATS])
def test_formats(f: dict[str, Any], python: str, tmp_path: Path) -> None:
    entry = next(e for e in MANIFEST if e["golden"] == f["golden"])
    golden = (ROOT / f["golden"]).read_text(encoding="utf-8")
    yaml = (ROOT / f["yaml"]).read_text(encoding="utf-8")
    r = run(python, [*args_of(entry, tmp_path), "--format", "yaml"])
    assert r.returncode == 0 and r.stdout == yaml, r.stderr
    assert to_json(read_yaml(yaml)) + "\n" == golden
    for n, file in f["indent"].items():
        want = (ROOT / file).read_text(encoding="utf-8")
        r = run(python, [*args_of(entry, tmp_path), "--indent", n])
        assert r.returncode == 0 and r.stdout == want, r.stderr
        assert to_json(read_json(want)) + "\n" == golden


def twin(spec: str) -> str:
    return re.sub(r"=tests/golden/inputs/(.*)\.json$", r"=tests/render/inputs/\1.yaml", spec)


TWINS = [e for e in MANIFEST if e.get("inputs") and [twin(s) for s in e["inputs"]] != e["inputs"]]


@pytest.mark.parametrize("entry", TWINS, ids=[e["golden"] for e in TWINS])
def test_yaml_twins(entry: dict[str, Any], python: str, tmp_path: Path) -> None:
    r = run(python, args_of(entry, tmp_path, [twin(s) for s in entry["inputs"]]))
    expected = (ROOT / entry["golden"]).read_text(encoding="utf-8")
    rejected = entry.get("rejected", False)
    assert r.returncode == (1 if rejected else 0), r.stderr
    assert (r.stderr if rejected else r.stdout) == expected


@pytest.mark.parametrize("case", INVALID, ids=[c["file"] for c in INVALID])
def test_invalid_documents(case: dict[str, Any], python: str) -> None:
    file = f"tests/render/invalid/{case['file']}"
    with pytest.raises(YamlError) as info:
        read_yaml((ROOT / file).read_text(encoding="utf-8"))
    assert str(info.value) == case["message"]
    r = run(python, ["validate", "tests/render/invalid/doc.decl", "--input", f"doc={file}"])
    want = (
        "tests/render/invalid/doc.decl: error [E6004] at doc: "
        f"bound document is not well-formed YAML: {file}: {case['message']}\n"
    )
    assert r.returncode == 1 and r.stderr == want

File: decl-rs/tests/render_test.rs
//! The render corpus (tests/render) through the native command line: the
//! format goldens of formats.json (`--format yaml`, `--indent n`, and the
//! YAML read back to the golden document), every golden document bound
//! from its YAML twin under inputs/, and the documents under invalid/ that
//! the reader must refuse with their messages (tests/render/README.md).
mod common;
use common::*;
use decl_lang::yaml::{read_yaml, to_json};
use std::process::Command;

fn run(root: &std::path::Path, args: &[String]) -> (i32, String, String) {
    let out = Command::new(env!("CARGO_BIN_EXE_decl"))
        .args(args)
        .current_dir(root)
        .output()
        .unwrap();
    (
        out.status.code().unwrap_or(-1),
        String::from_utf8_lossy(&out.stdout).to_string(),
        String::from_utf8_lossy(&out.stderr).to_string(),
    )
}

// a manifest entry's command line, with its inputs replaced when asked
fn args_of(
    root: &std::path::Path,
    e: &Value,
    tmp: &std::path::Path,
    inputs: Option<Vec<String>>,
) -> Vec<String> {
    let rejected = matches!(get(e, "rejected"), Some(Value::Bool(true)));
    let module = match get(e, "module") {
        Some(m) => text(m).to_string(),
        None => {
            let blocks = regex::Regex::new(r"(?s)```decl\n(.*?)```").unwrap();
            let md = std::fs::read_to_string(root.join(text(get(e, "markdown").unwrap()))).unwrap();
            let src: Vec<&str> = blocks
                .captures_iter(&md)
                .map(|c| c.get(1).unwrap().as_str())
                .collect();
            let p = tmp.join("guide.decl");
            std::fs::write(&p, src.join("\n")).unwrap();
            p.to_string_lossy().to_string()
        }
    };
    let mut args: Vec<String> = vec![
        if rejected { "validate" } else { "evaluate" }.into(),
        module,
    ];
    let specs: Vec<String> = match inputs {
        Some(v) => v,
        None => match get(e, "inputs") {
            Some(Value::JArr(items)) => items.iter().map(|s| text(s).to_string()).collect(),
            _ => vec![],
        },
    };
    for spec in specs {
        args.push("--input".into());
        args.push(spec);
    }
    if let Some(o) = get(e, "output") {
        args.push("--output".into());
        args.push(text(o).into());
    }
    args
}

fn manifest(root: &std::path::Path) -> Vec<Value> {
    let m = parse(&std::fs::read_to_string(root.join("tests/golden/manifest.json")).unwrap());
    let Value::JArr(entries) = m else {
        panic!("a list")
    };
    entries.iter().cloned().collect()
}

#[test]
fn formats() {
    let root = root();
    let tmp = temp_dir("render");
    let entries = manifest(&root);
    let formats = parse(&std::fs::read_to_string(root.join("tests/render/formats.json")).unwrap());
    let Value::JArr(formats) = formats else {
        panic!("a list")
    };
    let mut failures = vec![];
    for f in formats.iter() {
        let golden_path = text(get(f, "golden").unwrap());
        let e = entries
            .iter()
            .find(|e| text(get(e, "golden").unwrap()) == golden_path)
            .expect("a manifest entry");
        let golden = std::fs::read_to_string(root.join(golden_path)).unwrap();
        let yaml_path = text(get(f, "yaml").unwrap());
        let yaml = std::fs::read_to_string(root.join(yaml_path)).unwrap();
        let mut args = args_of(&root, e, &tmp, None);
        args.push("--format".into());
        args.push("yaml".into());
        let (code, out, _) = run(&root, &args);
        if code != 0 || out != yaml {
            failures.push(format!(
                "{yaml_path}: --format yaml: exit {code}, {}",
                first_diff(&yaml, &out)
            ));
        }
        let back = to_json(&read_yaml(&yaml).unwrap(), 0) + "\n";
        if back != golden {
            failures.push(format!(
                "{yaml_path}: read back: {}",
                first_diff(&golden, &back)
            ));
        }
        if let Some(Value::JObj(indents)) = get(f, "indent") {
            for (n, file) in indents.iter() {
                let want = std::fs::read_to_string(root.join(text(file))).unwrap();
                let mut args = args_of(&root, e, &tmp, None);
                args.push("--indent".into());
                args.push(n.clone());
                let (code, out, _) = run(&root, &args);
                if code != 0 || out != want {
                    failures.push(format!(
                        "{}: --indent {n}: exit {code}, {}",
                        text(file),
                        first_diff(&want, &out)
                    ));
                }
                let parsed = to_json(&read_json(&want).ok().expect("JSON"), 0) + "\n";
                if parsed != golden {
                    failures.push(format!(
                        "{}: parses back: {}",
                        text(file),
                        first_diff(&golden, &parsed)
                    ));
                }
            }
        }
    }
    let _ = std::fs::remove_dir_all(&tmp);
    assert!(
        failures.is_empty(),
        "format failures:\n  {}",
        failures.join("\n  ")
    );
}

#[test]
fn yaml_twins() {
    let root = root();
    let tmp = temp_dir("render");
    let re = regex::Regex::new(r"=tests/golden/inputs/(.*)\.json$").unwrap();
    let mut failures = vec![];
    for e in manifest(&root) {
        let Some(Value::JArr(inputs)) = get(&e, "inputs") else {
            continue;
        };
        let twins: Vec<String> = inputs
            .iter()
            .map(|s| {
                re.replace(text(s), "=tests/render/inputs/$1.yaml")
                    .to_string()
            })
            .collect();
        if twins.iter().zip(inputs.iter()).all(|(t, s)| t == text(s)) {
            continue;
        }
        let rejected = matches!(get(&e, "rejected"), Some(Value::Bool(true)));
        let args = args_of(&root, &e, &tmp, Some(twins.clone()));
        let (code, out, err) = run(&root, &args);
        let expected =
            std::fs::read_to_string(root.join(text(get(&e, "golden").unwrap()))).unwrap();
        let got = if rejected { err } else { out };
        if code != if rejected { 1 } else { 0 } || got != expected {
            failures.push(format!(
                "{}: exit {code}, {}",
                text(get(&e, "golden").unwrap()),
                first_diff(&expected, &got)
            ));
        }
    }
    let _ = std::fs::remove_dir_all(&tmp);
    assert!(
        failures.is_empty(),
        "twin failures:\n  {}",
        failures.join("\n  ")
    );
}

#[test]
fn invalid_documents() {
    let root = root();
    let cases =
        parse(&std::fs::read_to_string(root.join("tests/render/invalid/cases.json")).unwrap());
    let Value::JArr(cases) = cases else {
        panic!("a list")
    };
    let mut failures = vec![];
    for c in cases.iter() {
        let file = format!("tests/render/invalid/{}", text(get(c, "file").unwrap()));
        let message = text(get(c, "message").unwrap());
        let reader = match read_yaml(&std::fs::read_to_string(root.join(&file)).unwrap()) {
            Ok(_) => String::new(),
            Err(e) => e.message(),
        };
        if reader != message {
            failures.push(format!(
                "{file}: the reader says {reader:?}, not {message:?}"
            ));
        }
        let args: Vec<String> = vec![
            "validate".into(),
            "tests/render/invalid/doc.decl".into(),
            "--input".into(),
            format!("doc={file}"),
        ];
        let (code, _, err) = run(&root, &args);
        let want = format!("tests/render/invalid/doc.decl: error [E6004] at doc: bound document is not well-formed YAML: {file}: {message}\n");
        if code != 1 || err != want {
            failures.push(format!("{file}: exit {code}, {}", first_diff(&want, &err)));
        }
    }
    assert!(
        failures.is_empty(),
        "invalid-document failures:\n  {}",
        failures.join("\n  ")
    );
}
